### PR TITLE
feat(dart): use first version of constraint for dependencies using SDK version

### LINKED ISSUE #6017: feat(dart): use minimum SDK version for packages that use SDK for package version
## Description
Some dart packages use `SDK` version as package versions.
But `pubspec.lock` and `pubspec.yaml` files contain version range for `SDK`.
More details here - https://github.com/aquasecurity/trivy/discussions/5984#discussioncomment-8229400

```[tasklist]
### Tasks
- [x] use minimum SDK version for packages that use SDK for package version
- [x] add log message about using minimum SDK version from range
```

### Discussed in https://github.com/aquasecurity/trivy/discussions/5984

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
Use first version of constraint for dependencies using SDK version.
e.g. in this case the version of `flutter_test` should be `3.3.0`:
```yaml
  flutter_test:
    dependency: "direct dev"
    description: flutter
    source: sdk
    version: "0.0.0"
sdks:
  dart: ">=2.18.0 <3.0.0"
  flutter: "^3.3.0"
```

More information in #5984 

## Related issues
- Close #6017

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/dependency/parser/dart/pub/parse.go [hunk 1]
```diff
@@ -23,30 +24,50 @@ func NewParser() types.Parser {
...
+		}
+
 		// We would like to exclude dev dependencies, but we cannot identify
 		// which indirect dependencies were introduced by dev dependencies
 		// as there are 3 dependency types, "direct main", "direct dev" and "transitive".
 		// It will be confusing if we exclude direct dev dependencies and include transitive dev dependencies.
 		// We decided to keep all dev dependencies until Pub will add support for "transitive main" and "transitive dev".
 		lib := types.Library{
-			ID:       pkgID(name, dep.Version),
+			ID:       utils.PackageID(name, dep.Version),
```
Comment (knqyf263) [inline pkg/dependency/parser/dart/pub/parse.go:68]:
```suggestion
			ID:       utils.PackageID(name, version),
```

Comment (DmitriyLewen) [reply pkg/dependency/parser/dart/pub/parse.go]: Thanks that you found my typo!
Fixed in https://github.com/aquasecurity/trivy/pull/6239/commits/c5ce4bc138969265cca1fd58f00664fa645e827d

File: pkg/dependency/parser/dart/pub/parse.go [hunk 2]
```diff
@@ -64,3 +91,30 @@ func (p Parser) relationship(dep string) ftypes.Relationship {
...
+	switch {
+	case strings.HasPrefix(constraint, ">="):
+		constraint = strings.TrimPrefix(constraint, ">=")
+		constraint, _, _ = strings.Cut(constraint, " ")
+		return constraint
+	case strings.HasPrefix(constraint, "^"):
+		return strings.TrimPrefix(constraint, "^")
+	}
+	return ""
+}
```
Comment (knqyf263) [inline pkg/dependency/parser/dart/pub/parse.go:107]: I think we want to parse the constraints in a more stable way. What if we export `constraint` and add methods like `func (c *Constraint) Version() string` and `func (c *Constraint) Operator() string`?
https://github.com/aquasecurity/go-version/blob/637058cfe4921e395b56f195224538d7eac61520/pkg/version/constraint.go#L55-L59

```
type Constraint struct {
	version  Version
	operator string // e.g. "=", ">=", "^"
	operatorFunc operatorFunc
	original string
}

func (c *Constraint) Version() string {
    return c.version
}

func (c *Constraint) Operator() string {
    return c.operator
}
```

Comment (knqyf263) [reply pkg/dependency/parser/dart/pub/parse.go]: Merged
https://github.com/aquasecurity/go-version/pull/6

Comment (DmitriyLewen) [reply pkg/dependency/parser/dart/pub/parse.go]: Updated this PR -https://github.com/aquasecurity/trivy/pull/6239/commits/91c2b1b63aa0b7f607ef97a0db5ed3e42c43c32e

File: pkg/dependency/parser/dart/pub/parse.go [hunk 3]
```diff
@@ -64,3 +93,34 @@ func (p Parser) relationship(dep string) ftypes.Relationship {
...
+	css, err := goversion.NewConstraints(constraint)
+	if err != nil {
+		return "", xerrors.Errorf("unable to parse constraints: %w", err)
+	}
+
+	// Dart uses only `>=` and `^` operators:
+	// cf. https://dart.dev/tools/pub/dependencies#traditional-syntax
+	constraints := css.List()
+	// We only need to get the first version from the range
+	if constraints[0][0].Operator() != ">=" && constraints[0][0].Operator() != "^" {
```
Comment (knqyf263) [inline pkg/dependency/parser/dart/pub/parse.go:108]: Can we check the length so it won't panic? I presume `css.List()` doesn't return empty, but unexpected things always happen. It's better to ensure it's non-empty.

Comment (DmitriyLewen) [reply pkg/dependency/parser/dart/pub/parse.go:111]: it won't hurt.
Added check in https://github.com/aquasecurity/trivy/pull/6239/commits/35a6f0ac586d0d759f36bccf53612ac75ee9d6d6

File: pkg/dependency/parser/dart/pub/parse.go [hunk 4]
```diff
@@ -16,37 +18,64 @@ const (
...
+		// In this case dep.Version == `0.0.0`.
+		// We can't get versions for these dependencies.
+		// Therefore, we use the first version of the SDK constraint specified in the Description.
+		// See https://github.com/aquasecurity/trivy/issues/6017
+		version := dep.Version
+		if version == "0.0.0" && dep.Source == "sdk" {
+			if constraint, ok := l.Sdks[string(dep.Description)]; ok {
+				v, err := firstVersionOfConstrain(constraint)
+				if err != nil {
+					p.logger.Warn("unable to get sdk version from constraint: %w", log.Err(err))
```
Comment (knqyf263) [inline pkg/dependency/parser/dart/pub/parse.go]:
```suggestion
					p.logger.Warn("Unable to get sdk version from constraint", log.Err(err))
```

**Review by knqyf263 — APPROVED** [review]

File: pkg/dependency/parser/dart/pub/parse.go (hunk 4 above)
Comment (knqyf263) [reply pkg/dependency/parser/dart/pub/parse.go]: Fixed in https://github.com/aquasecurity/trivy/pull/6239/commits/191270c0a3fa995e888f1a3d88c29307844d70b4

**Pull request merged**